src/server/index.ts

```diff
@@ -1,93 +1,43 @@
-            }
-        })
-    }
-
-    async serial_open(
-        payload: { port: string; options?: Options },
-        callback: (res: Response<string>) => void
-    ): Promise<void> {
-        this.ryder_serial = new RyderSerial(payload.port, payload.options)
-        new Promise<string>((resolve, reject) => {
-            if (!this.ryder_serial) {
-                reject("Ryder Serial does not exist for some reason")
-                return
-            }
-
-            this.ryder_serial.on("failed", (error: Error) => {
-                reject(
-                    `Could not connect to the Ryder on the specified port. Wrong port or it is currently in use: ${error}`
-                )
-                return
-            })
-            this.ryder_serial.on("open", async () => {
-                if (!this.ryder_serial) {
-                    reject("Ryder serial was destroyed")
-                    return
-                }
-                const response = await this.ryder_serial.send(RyderSerial.COMMAND_INFO)
-                const info = typeof response === "number" ? response.toString() : response
-                if (!info || info.substr(0, 5) !== "ryder") {
-                    reject(`Device at ${payload.port} does not appear to be a Ryder device`)
-                } else {
-                    resolve(info)
-                }
-                return
-            })
-            this.ryder_serial.on("wait_user_confirm", () => {
-                resolve("Confirm or cancel on Ryder device.")
-                return
-            })
-        })
-            .then((res: string) => callback({ data: res }))
-            .catch(error =>
-                callback({
-                    source: error,
-                    error: error,
-                })
-            )
-            .finally(() => this.ryder_serial?.close())
-
-        this.handler?.socket.emit("serial:opened")
-    }
+import WebSocket from 'ws';
+import config, { Config } from "../config";
+import { log } from './logger';
+import { IncomingMessage } from 'http';
+import EventEmitter from 'events';
+import { RyderRpcHost } from './ryder';
+import { AsyncRpcProvider, RpcRequest } from './asyncrpcprovider';
+
+
+export class BridgeServer extends EventEmitter {
+	protected rpc_provider: AsyncRpcProvider;
+	protected wss!: WebSocket.Server;
+
+	constructor(rpc_provider: AsyncRpcProvider) {
+		super();
+		this.rpc_provider = rpc_provider;
+	}
+
+	listen(config: Config) {
+		this.wss = new WebSocket.Server({ host: config.host, port: config.port });
+		this.emit('listening');
+		this.wss.on('connection', this.handle_connection.bind(this));
+	}
+
+	protected handle_connection(client: WebSocket, request: IncomingMessage) {
+		if (!['127.0.0.1', '::1'].includes(request.socket.remoteAddress!))
+			return client.close(1);
+		this.emit('connection', client);
+		client.on('message', this.handle_message.bind(this));
+	}
+
+	protected handle_message(client: WebSocket, data: WebSocket.Data) {
```

Comment:
`type WebSocket.Data = string | Buffer | ArrayBuffer | Buffer[]`

are you returning out with the `if (typeof data !== "string") return;` b/c we don't want to handle `Buffer | ArrayBuffer | Buffer[]`? And if so, should we just make `data` parameter `string`?

Reply:
It should always be a string if we're sending JSON. Since it's JSON-RPC over ws I figured to reject the others immediately because it means the sender is doing something funny. Still, we can just read the buffers as UTF-8 strings and and parse them as JSON. What do you think?

Reply:
gotchya that makes sense — let's leave it for now but maybe later incorporate some sort of error message.

Reply:
Yeah sending a JSON-RPC error object makes sense.